clang/tools/checked-c-convert/ArrayBoundsInferenceConsumer.cpp

```diff
@@ -1,127 +1,558 @@
-        }
-      }
-  }
-  return false;
-}
+bool LocalVarABVisitor::VisitSwitchStmt(SwitchStmt *S) {
+  VarDecl *CondVar = S->getConditionVariable();
 
-// check if expression is a simple local variable
-// i.e., ptr = .
-// if yes, return the referenced local variable as the return
-// value of the argument.
-bool LocalVarABVisitor::isExpressionSimpleLocalVar(Expr *toCheck, Decl **targetDecl) {
-  if (DeclRefExpr *DRE = dyn_cast<DeclRefExpr>(toCheck))
-    if (DeclaratorDecl *FD = dyn_cast<DeclaratorDecl>(DRE->getDecl()))
-      if (Decl *V = dyn_cast<Decl>(FD)) {
-        *targetDecl = V;
-        return true;
-      }
-  return false;
+  // If this is a parameter declaration? Then this parameter cannot be length.
+  if (CondVar != nullptr)
+    if (ParmVarDecl *PD = dyn_cast<ParmVarDecl>(CondVar))
+      NonLengthParameters.insert(PD);
 }
 
-void LocalVarABVisitor::dumpNotArrayIdentifiedVariable(Decl *LHS, Expr *RHS, raw_ostream &O) {
-#ifdef DEBUG
-  O << "Not identified as a array variable.\n RHS:";
-  RHS->dump(O);
-  O << "\n LHS:";
-  LHS->dump(O);
-  O << "\n";
-#endif
+// Check if the provided parameter cannot be a length of an array.
+bool LocalVarABVisitor::isNonLengthParameter(ParmVarDecl* PVD) {
+  if (PVD->getType().getTypePtr()->isEnumeralType())
+    return true;
+  return NonLengthParameters.find(PVD) != NonLengthParameters.end();
 }
 
-Expr *LocalVarABVisitor::removeAuxiliaryCasts(Expr *srcExpr) {
-  srcExpr = srcExpr->IgnoreParenImpCasts();
-  if (CStyleCastExpr *C = dyn_cast<CStyleCastExpr>(srcExpr))
-    srcExpr = C->getSubExpr();
-  srcExpr = srcExpr->IgnoreParenImpCasts();
-  return srcExpr;
-}
+void AddArrayHeuristics(ASTContext *C, ProgramInfo &I, FunctionDecl *FD) {
+  if (FD->isThisDeclarationADefinition() && FD->hasBody()) {
+    // Heuristic: If the function has just a single parameter
+    // and we found that it is an array then it must be an Nt_array.
+    const Type *Ty = FD->getTypeSourceInfo()->getTypeLoc().getTypePtr();
+    const FunctionProtoType *FT = Ty->getAs<FunctionProtoType>();
+    if (FT != nullptr) {
+      if (FT->getNumParams() == 1) {
+        ParmVarDecl *PVD = FD->getParamDecl(0);
+        auto &CS = I.getConstraints();
+        std::set<ConstraintVariable *> defsCVar = I.getVariable(PVD, C, true);
+        for (auto constraintVar: defsCVar)
+          if (PVConstraint *PV = dyn_cast<PVConstraint>(constraintVar)) {
+            auto &Cvars = PV->getCvars();
+            if (Cvars.size() > 0) {
+              // We should constraint only the outer most constraint variable.
+              auto CVar = *(Cvars.begin());
+              CS.getOrCreateVar(CVar)->setNtArrayIfArray();
+            }
+          }
+      } else if (FD->getNameInfo().getAsString() == std::string("main") &&
+                 FT->getNumParams() == 2) {
+        // If the function is `main` then we know second argument is _Array_ptr.
+        ParmVarDecl *Argv = FD->getParamDecl(1);
+        assert(Argv != NULL);
```

Comment:
Please use nullptr instead of NULL. It's also good practice to give a short descriptive message for an assert.